t/Controller/Jobset/constraints.t

```diff
@@ -0,0 +1,165 @@
+use strict;
+use Setup;
+my %ctx = test_init();
+
+require Hydra::Schema;
+require Hydra::Model::DB;
+require Hydra::Helper::Nix;
+
+use Test2::V0;
+require Catalyst::Test;
+use HTTP::Request::Common;
+Catalyst::Test->import('Hydra');
+
+my $db = Hydra::Model::DB->new;
+hydra_setup($db);
+
+# This test checks a matrix of jobset configuration options for constraint violations.
+
+# Modified from https://github.com/szabgab/slides/blob/408b3712b6bc929d4ea9f382a0cc930c2bc3c5fb/perl/examples/references/compare_hashes.pl
+sub compare_hashes {
+    my ($first, $second) = @_;
+    my $ret = 0;
+
+    foreach my $k (keys %{ $first }) {
+        if (not exists $second->{$k}) {
+            $ret = 1;
+        } elsif (not defined $first->{$k} and not defined $second->{$k}) {
+        } elsif (defined $first->{$k} and not defined $second->{$k}) {
+            $ret = 1;
+        } elsif (not defined $first->{$k} and defined $second->{$k}) {
+            $ret = 1;
+        } elsif ($first->{$k} ne $second->{$k}) {
+            $ret = 1;
+        }
+    }
+
+    foreach my $k (keys %{ $second }) {
+        if (not exists $first->{$k}) {
+            $ret = 1;
+        }
+    }
+
+    return $ret;
+};
+
+my @jobset_types = ( 0, 1, 2 );
+my @nixexprinputs = ( undef, "input" );
+my @nixexprpaths = ( undef, "path" );
+my @flakes = ( undef, "flake" );
+
+my @expected_failing;
+my @expected_succeeding = (
+    {
+        "jobset_type" => 0,
+        "nixexprinput" => "input",
+        "nixexprpath" => "path",
+        "flake" => undef,
+    },
+    {
+        "jobset_type" => 1,
+        "nixexprinput" => undef,
+        "nixexprpath" => undef,
+        "flake" => "flake",
+    }
+);
+
+# Construct a matrix of parameters that should violate the Jobsets table's constraints.
+foreach my $jobset_type (@jobset_types) {
+    foreach my $nixexprinput (@nixexprinputs) {
+        foreach my $nixexprpath (@nixexprpaths) {
+            foreach my $flake (@flakes) {
+                my $hash = {
+                    "jobset_type" => $jobset_type,
+                    "nixexprinput" => $nixexprinput,
+                    "nixexprpath" => $nixexprpath,
+                    "flake" => $flake,
+                };
+
+                push(@expected_failing, $hash)
+                    if (compare_hashes($hash, $expected_succeeding[0]) && compare_hashes($hash, $expected_succeeding[1]));
```

Comment:
How about turning this in to a foreach loop, so if there are additional cases later this bit of code doesn't need to be updated?